crates/hyperdrive-math/src/lp/remove.rs

```diff
@@ -1,6 +1,312 @@
+                current_block_timestamp,
+                active_lp_total_supply,
+                withdrawal_shares_total_supply,
+                SHARE_PROCEEDS_MAX_ITERATIONS,
+            )?;
+
+        // Update withdrawal pool after distributing excess idle.
+        let ready_to_withdraw =
+            self.withdrawal_shares_ready_to_withdraw() + withdrawal_shares_redeemed;
+        let withdrawal_share_proceeds = self.withdrawal_shares_proceeds() + share_proceeds;
+
+        // Clamp the shares to the total amount of shares ready for withdrawal
+        // to avoid unnecessary reverts. We exit early if the user has no shares
+        // available to redeem.
+        let mut withdrawal_shares_redeemed = withdrawal_shares;
+        if withdrawal_shares_redeemed > ready_to_withdraw {
+            withdrawal_shares_redeemed = ready_to_withdraw;
+        }
+        if withdrawal_shares_redeemed == fixed!(0) {
+            return Ok((fixed!(0), fixed!(0)));
+        }
+
+        // NOTE: Round down to underestimate the share proceeds.
+        //
+        // The LP gets the pro-rata amount of the collected proceeds.
+        let share_proceeds =
+            withdrawal_shares_redeemed.mul_div_down(withdrawal_share_proceeds, ready_to_withdraw);
+
+        // Withdraw the share proceeds to the user
+        let proceeds = self.withdraw(share_proceeds, vault_share_price)?;
+
+        // NOTE: Round up to make the check more conservative.
+        //
+        // Ensure proceeds meet minimum output per share
+        if proceeds < min_output_per_share.mul_up(withdrawal_shares_redeemed) {
+            return Err(eyre!("Output limit not met"));
+        }
+
+        Ok((proceeds, withdrawal_shares_redeemed))
+    }
+
+    /// Distribute as much of the excess idle as possible to the withdrawal
+    /// pool while holding the LP share price constant.
+    fn calculate_distribute_excess_idle_safe(
+        &self,
+        current_block_timestamp: U256,
+        withdrawal_shares_total_supply: FixedPoint,
+        vault_share_price: FixedPoint,
+        max_iterations: u64,
+    ) -> Result<(FixedPoint, FixedPoint, bool)> {
+        let withdrawal_shares_total_supply =
+            withdrawal_shares_total_supply - self.withdrawal_shares_ready_to_withdraw();
+
+        // If there are no withdrawal shares, then there is nothing to
+        // distribute.
+        if withdrawal_shares_total_supply == fixed!(0) {
+            return Ok((fixed!(0), fixed!(0), true));
+        }
+
+        // If there is no excess idle, then there is nothing to distribute.
+        let idle = self.calculate_idle_share_reserves();
+        if idle == fixed!(0) {
+            return Ok((fixed!(0), fixed!(0), true));
+        }
+
+        // Calculate the amount of withdrawal shares that should be redeemed
+        // and their share proceeds.
+        let (withdrawal_shares_redeemed, share_proceeds) = self.calculate_distribute_excess_idle(
```

Comment:
solidity has `_distributeExcessIdleSafe()`, which I believe does the actual distribution & includes checks, and `calculateDistributeExcessIdle()` to compute the `withdrawalSharesRedeemed` and `shareProceeds`.

Here in rust you're renaming the first function to `calculate_distribute_excess_idle_safe` and the second to `calculate_distribute_excess_idle`. I think these names are confusing -- it implies that they're the same function and one is the "safe" version of it. 

This is a case where we should depart from exactly copying solidity in favor of rust readability. I think you should rename one of these functions.

Reply:
> This is a case where we should depart from exactly copying solidity in favor of rust readability. I think you should rename one of these functions.

This was my feeling as well, but honestly it's too much of a pain to maintain two different implementations for too little gain.  I tore my hair out trying to make things more readable only to revert to a more mirrored approach.  

I can rename this to whatever we want though, it's 'safe' because of the extra checks, how about renaming to get rid of the 'calculate' part and return updated `State` along with the updated values?